### PR TITLE
DB-8901 Consistency check in SpliceController.batchFetch

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java [hunk 1]
```diff
@@ -168,13 +168,15 @@ public boolean batchFetch(List<RowLocation> locations, List<ExecRow> destRows, F
         return batchFetch(locations,destRows,validColumns,false);
     }
     @Override
-    public boolean batchFetch(List<RowLocation> locations, List<ExecRow> destRows,FormatableBitSet validColumns,boolean waitForLock) throws StandardException{
-        if (locations.isEmpty())
+    public boolean batchFetch(List<RowLocation> locations, List<ExecRow> destRows, FormatableBitSet validColumns, boolean waitForLock) throws StandardException{
+        if (locations.size() != destRows.size())
             return false;
```
Comment (msirek) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: getDescriptorViaIndexMinion calls batchFetch.  So, if there's a list size mismatch, getDescriptorViaIndexMinion might not find any tuple descriptors.  Would this error situation always be caught or should we change the 'return false' here into an assertion?

File: splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java [hunk 2]
```diff
@@ -185,18 +187,19 @@ public boolean batchFetch(List<RowLocation> locations, List<ExecRow> destRows,Fo
...
                 row.setRowArray(destRow);
                 row.resetRowArray();
-                DataCell keyValue=result.userData();
-                rowDecoder.set(keyValue.valueArray(),keyValue.valueOffset(),keyValue.valueLength());
+                DataCell keyValue = result.userData();
+                rowDecoder.set(keyValue.valueArray(), keyValue.valueOffset(), keyValue.valueLength());
                 rowDecoder.decode(row);
                 i++;
             }
+            assert i == locations.size();
```
Comment (msirek) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: I believe assertions only get thrown on standalone.  If you want the error to be caught on a cluster as well, maybe an Exception would be better (I don't know this for sure about assertions, but that's the behavior I recall seeing.  You may want to test it).

Comment (jyuanca) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: Assertion is not enabled when splice is deployed on a cluster

Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: This is development specific and should not happen on a production machine.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java (hunk 1 above)
Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: That's a good point, I changed the way we catch error in getDescriptorViaIndexMinion to reflect what's happening in here.

I think that there should be some more refactoring in the future in there. (legacy assertions, ..)

Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/store/access/base/SpliceController.java]: That's a good point, I changed the way we catch error in getDescriptorViaIndexMinion to reflect what's happening in here.

I think that there should be some more refactoring in the future in there. (legacy assertions, ..)

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java [hunk 3]
```diff
@@ -7183,8 +7183,7 @@ public final TupleDescriptor getDescriptorViaIndex(int indexId,
             rowLocations.add((RowLocation) indexRow1.getColumn(
                     indexRow1.nColumns()).cloneValue(true));
         }
-        heapCC.batchFetch(rowLocations,outRows,null);
-        if (outRows.size() != i) {
+        if (!heapCC.batchFetch(rowLocations,outRows,null)) {
             if(SanityManager.DEBUG){
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java]: It'd be better to substitute this SanityManager block for something else that raises an error

**Comment by jyuanca** [comment]
@arnaud-splice Can you please resolve change request?

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**